### ISSUE TITLE
Installation error: "ambiguous overloaded method"

### ISSUE THREAD

**Issue by kcgthb** [qutip/qutip#723]
Hi!

I'd like to report an installation issue on CentOS/RHEL platforms. I've tested various Python versions (2.7, 3.4) on both RHEL 6.x and 7.x, and the error is consistent, and happens during compilation of some Cython files.

The error is: `ambiguous overloaded method` referring to the `abs()` function, and happens in two different places.

Full installation output:
```
$ pip install qutip
Collecting qutip
  Using cached qutip-4.1.0.tar.gz
    Complete output from command python setup.py egg_info:
    
    Error compiling Cython file:
    ------------------------------------------------------------
    ...
        cdef double * col_sum = <double *>PyDataMem_NEW_ZEROED(ncols, sizeof(double))
        cdef double max_col = 0
        for ii in range(nrows):
            for jj in range(ptr[ii], ptr[ii+1]):
                k = ind[jj]
                col_sum[k] += abs(data[jj])
                                ^
    ------------------------------------------------------------
    
    qutip/cy/sparse_utils.pyx:316:29: ambiguous overloaded method
    
    Error compiling Cython file:
    ------------------------------------------------------------
    ...
        cdef size_t ii, jj
        cdef double * row_sum = <double *>PyDataMem_NEW_ZEROED(nrows, sizeof(double))
        cdef double max_row = 0
        for ii in range(nrows):
            for jj in range(ptr[ii], ptr[ii+1]):
                row_sum[ii] += abs(data[jj])
                                 ^
    ------------------------------------------------------------
    
    qutip/cy/sparse_utils.pyx:335:30: ambiguous overloaded method
    Compiling qutip/cy/spmatfuncs.pyx because it depends on /home/kilian/qutip/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd.
    Compiling qutip/cy/stochastic.pyx because it depends on /home/kilian/qutip/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd.
    Compiling qutip/cy/sparse_utils.pyx because it depends on /home/kilian/qutip/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd.
    Compiling qutip/cy/graph_utils.pyx because it depends on /home/kilian/qutip/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd.
    Compiling qutip/cy/interpolate.pyx because it depends on /home/kilian/qutip/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd.
    Compiling qutip/cy/spmath.pyx because it depends on /home/kilian/qutip/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd.
    Compiling qutip/cy/heom.pyx because it depends on /home/kilian/qutip/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd.
    Compiling qutip/cy/math.pyx because it depends on /home/kilian/qutip/lib/python2.7/site-packages/Cython/Includes/libc/math.pxd.
    Compiling qutip/cy/spconvert.pyx because it depends on /home/kilian/qutip/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd.
    Compiling qutip/cy/ptrace.pyx because it depends on /home/kilian/qutip/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd.
    Compiling qutip/cy/testing.pyx because it depends on /home/kilian/qutip/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd.
    Compiling qutip/control/cy_grape.pyx because it depends on /home/kilian/qutip/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd.
    [ 1/12] Cythonizing qutip/control/cy_grape.pyx
    [ 2/12] Cythonizing qutip/cy/graph_utils.pyx
    [ 3/12] Cythonizing qutip/cy/heom.pyx
    [ 4/12] Cythonizing qutip/cy/interpolate.pyx
    [ 5/12] Cythonizing qutip/cy/math.pyx
    [ 6/12] Cythonizing qutip/cy/ptrace.pyx
    [ 7/12] Cythonizing qutip/cy/sparse_utils.pyx
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/state/partition1/local-scratch/kilian/16342884/pip-build-7c5rN_/qutip/setup.py", line 216, in <module>
        ext_modules = cythonize(EXT_MODULES),
      File "/home/kilian/qutip/lib/python2.7/site-packages/Cython/Build/Dependencies.py", line 1016, in cythonize
        cythonize_one(*args)
      File "/home/kilian/qutip/lib/python2.7/site-packages/Cython/Build/Dependencies.py", line 1138, in cythonize_one
        raise CompileError(None, pyx_file)
    Cython.Compiler.Errors.CompileError: qutip/cy/sparse_utils.pyx
```

Tested with:
* Cython 0.26
* numpy 1.13.1
* gcc 4.9.1

Any recommendation on how to fix this? 

Thanks!

**Comment by nonhermitian** [comment]
Did it work with the previous Cython 0.25?

**Comment by kcgthb** [comment]
I don't know, that's my first `qutip` installation. :)

But I just tried, and it indeed works with Cython 0.25, thanks for the pointer.

**Comment by nonhermitian** [comment]
The latest version of QuTiP (4.2) should work just fine.  There was some importing slight of hand in the old (4.1) file that was causing the issue.

**Comment by nonhermitian** [comment]
Please let me know if the latest version works, so that I may close this issue.

**Comment by kcgthb** [comment]
Hi @nonhermitian,

I don't see any 4.2 release nor v4.2 tag, which commit should I test?

**Comment by nonhermitian** [comment]
You can just grab the master.

**Comment by kcgthb** [comment]
So I did just that, and the "overloaded method" error doesn't happen again, which is good.
But installation still fails on other errors, I believe related to NumPy:

```
./setup.py install --user
Compiling qutip/cy/spmatfuncs.pyx because it changed.
Compiling qutip/cy/stochastic.pyx because it changed.
Compiling qutip/cy/sparse_utils.pyx because it changed.
Compiling qutip/cy/graph_utils.pyx because it changed.
Compiling qutip/cy/interpolate.pyx because it changed.
Compiling qutip/cy/spmath.pyx because it changed.
Compiling qutip/cy/heom.pyx because it changed.
Compiling qutip/cy/math.pyx because it changed.
Compiling qutip/cy/spconvert.pyx because it changed.
Compiling qutip/cy/ptrace.pyx because it changed.
Compiling qutip/cy/testing.pyx because it changed.
Compiling qutip/cy/brtools.pyx because it changed.
Compiling qutip/cy/brtools_testing.pyx because it changed.
[ 1/13] Cythonizing qutip/cy/brtools.pyx

Error compiling Cython file:
------------------------------------------------------------
...

DTYPE = np.float64
         ^
------------------------------------------------------------

qutip/cy/parameters.pxi:2:10: cimported module has no attribute 'float64'

Error compiling Cython file:
------------------------------------------------------------
...

DTYPE = np.float64
ctypedef np.float64_t DTYPE_t

ITYPE = np.int32
         ^
------------------------------------------------------------

qutip/cy/parameters.pxi:5:10: cimported module has no attribute 'int32'

Error compiling Cython file:
------------------------------------------------------------
...
ctypedef np.float64_t DTYPE_t

ITYPE = np.int32
ctypedef np.int32_t ITYPE_t

CTYPE = np.complex128
         ^
------------------------------------------------------------

qutip/cy/parameters.pxi:8:10: cimported module has no attribute 'complex128'

Error compiling Cython file:
------------------------------------------------------------
...
ctypedef np.int32_t ITYPE_t

CTYPE = np.complex128
ctypedef np.complex128_t CTYPE_t

CTYPE = np.int64
         ^
------------------------------------------------------------

qutip/cy/parameters.pxi:11:10: cimported module has no attribute 'int64'
Traceback (most recent call last):
  File "./setup.py", line 233, in <module>
    ext_modules = cythonize(EXT_MODULES),
  File "/dev/shm/qutip/lib/python2.7/site-packages/Cython/Build/Dependencies.py", line 1016, in cythonize
    cythonize_one(*args)
  File "/dev/shm/qutip/lib/python2.7/site-packages/Cython/Build/Dependencies.py", line 1138, in cythonize_one
    raise CompileError(None, pyx_file)
Cython.Compiler.Errors.CompileError: qutip/cy/brtools.pyx
```

Versions:
```
Python 2.7.5 (default, Dec  3 2013, 08:35:16) 
[GCC 4.4.6 20120305 (Red Hat 4.4.6-4)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
>>> numpy.version.version
'1.13.1'
>>> import Cython
>>> print Cython.__version__
0.26
```

**Comment by sahmed95** [comment]
Hi, could you try with python 3.5+ ? It is [recommended](http://qutip.org/docs/4.1/installation.html). Also, if you have the package manager conda ([Miniconda](https://conda.io/miniconda.html)). Installing QuTiP 4.2 is as easy as `conda install -c conda-forge qutip=4.2`

**Comment by nonhermitian** [comment]
Try Pull #731.  It seems to fix the issues with Cython 0.26 for me.

**Comment by kcgthb** [comment]
I have the same exact same NumPy errors in a fresh VirtualEnv with Python 3.6.1 and #731.

**Comment by nonhermitian** [comment]
With what errors?

**Comment by nonhermitian** [comment]
Tested here at work on Win with Cython 0.26, and built fine.

**Comment by kcgthb** [comment]
> With what errors?
The ones from https://github.com/qutip/qutip/issues/723#issuecomment-317495769

**Comment by ajgpitch** [comment]
@sahmed95 do you fancy making a CentOS/RHEL VM to see if you can replicate this issue?

**Comment by ajgpitch** [comment]
@kcgthb have you tested this inside of a conda env on CentOS/RHEL? 
I ask because that is our supported platform.

**Comment by ajgpitch** [comment]
The cluster that I use is running on CentOS 6.7. As it is important to me that qutip runs on this, then I will give our 4.2 version a test on this.

**Comment by ajgpitch** [comment]
I created a CentOS VM - which I think may have been unnecessary and a bit time consuming
I installed Miniconda and the qutip required libraries. Initially this was cython 0.25.2
I installed qutip 4.2 (the version we created for the release) from source with no error.
I upgraded to cython 0.26, for which I had to use the conda-forge channel.
I tried to install qutip 4.2 (the version we created for the release) from source and found the errors reported by @kcgthb 
I installed qutip 4.2 (the version from #731 ) from source with no error.
I would say therefore that this issue is fixed by #731
